pkg/loadbalancer/experimental/sync.go

```diff
@@ -0,0 +1,321 @@
+	if cfg.LBLocalSyncFile == "" || !lbcfg.EnableExperimentalLB {
+		return nil
+	}
+
+	file := cfg.LBLocalSyncFile
+	isYAML := false
+	switch {
+	case strings.HasSuffix(file, ".yaml"), strings.HasSuffix(file, ".yml"):
+		isYAML = true
+	case strings.HasSuffix(file, ".json"):
+		isYAML = false
+	default:
+		return fmt.Errorf("could not deduce file format from %q, expected .json or .yaml suffix", file)
+	}
+	ls := localSynchronizer{
+		log:    log,
+		file:   file,
+		isYAML: isYAML,
+		w:      w,
+	}
+	jg.Add(job.OneShot(
+		"sync",
+		ls.syncLoop,
+		job.WithRetry(-1, &job.ExponentialBackoff{
+			Min: time.Minute,
+			Max: 5 * time.Minute,
+		}),
+	))
+	return nil
+}
+
+// localSynchronizer syncs Source=LocalAPI state from a file containing a
+// [structs.LoadBalancerState].
+type localSynchronizer struct {
+	log    *slog.Logger
+	file   string
+	isYAML bool
+	w      *Writer
+}
+
+func (s *localSynchronizer) syncLoop(ctx context.Context, health cell.Health) error {
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return err
+	}
+	defer watcher.Close()
+
+	// Watch the parent directory for changes and react to events on the specified
+	// file.
+	watchDir := "."
+	if path.IsAbs(s.file) {
+		watchDir = path.Dir(s.file)
+	}
+	if err := watcher.Add(watchDir); err != nil {
+		return fmt.Errorf("failed to watch %q: %w", s.file, err)
+	}
+
+	limiter := rate.NewLimiter(time.Second, 5)
+
+	for {
+		if state, err := s.load(); err != nil {
+			s.log.Error("Failed to synchronize", "file", s.file, "error", err)
+		} else {
+			txn := s.w.WriteTxn()
+			if err := s.synchronize(txn, state); err == nil {
+				txn.Commit()
+				// TODO summarize the changes
+				s.log.Info("Synchronized from local state", "file", s.file)
+			} else {
+				s.log.Error("Failed to synchronize", "file", s.file, "error", err)
+				txn.Abort()
+			}
+		}
+
+		// Rate limit the reprocessing of changes.
+		if err := limiter.Wait(ctx); err != nil {
+			return err
+		}
+
+	again:
+		select {
+		case ev := <-watcher.Events:
+			if path.Base(ev.Name) != path.Base(s.file) {
+				goto again
+			}
+			if ev.Op == fsnotify.Remove {
+				// Ignore removals. To clear out the local state an empty file is expected.
+				goto again
+			}
+		case <-ctx.Done():
+			return nil
+		}
+	}
+}
+
+func (s *localSynchronizer) load() (*structs.LoadBalancerState, error) {
+	b, err := os.ReadFile(s.file)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read file: %w", err)
+
+	}
+	var state structs.LoadBalancerState
+
+	// TODO: move format detection to constructor
+	if s.isYAML {
+		err = yaml.Unmarshal(b, &state)
+	} else {
+		err = json.Unmarshal(b, &state)
+	}
+
+	if err != nil {
+		return nil, err
+	}
+
+	return &state, nil
+
+}
+
+func (s *localSynchronizer) synchronize(txn WriteTxn, state *structs.LoadBalancerState) error {
+	// Delete all existing source=LocalAPI services/frontends/backends.
+	err := s.w.DeleteServicesBySource(txn, source.LocalAPI)
```

Comment:
Shouldn't we first upsert, and then delete no-longer existing svcs? Otherwise, we might introduce transient failures.

Reply:
I have similar question if this is really safe to do with active services.

Reply:
This is all transactional. The reconciler only sees the changes after the transaction is committed.